### PR TITLE
Goodness-of-fit diagnostics for Hawkes EM algorithm

### PR DESCRIPTION
**Pull request by claudio-tw**
This PR proposes to add new methods to the class `HawkesEM`, and to expand goodness-of-fit diagnostics for the expectation-maximisation algorithm.  The key capability is the computation of compensators of Hawkes processes, defined as time-integrals of the conditional intensity functions. (Notice that "compensator" is also used to refer to a subroutine in the calibration routine, but no confusion should arise).  
This PR also introduces google tests for the class `HawkesEM`.

### CODE REVIEW

**Review by Mbompr — APPROVED** [review]
Nice

**Comment by claudio-tw** [comment]
@PhilipDeegan Do you know why the build on Ubuntu is failing?

**Comment by PhilipDeegan** [comment]
consider rebasing https://github.com/X-DataInitiative/tick/pull/510, with that we can see if all the builds fail the same wya

File: lib/cpp-test/hawkes/inference/hawkes_em_gtest.cpp [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+TEST_F(HawkesEMTest, kernel_time_func_values) {
+  em.init_kernel_time_func(kernels);
+  std::vector<TimeFunction>& timefunc = em.get_kernel_time_func();
+  ASSERT_FALSE(timefunc.empty());
+  for (ulong u = 0; u < n_nodes; ++u) {
+    for (ulong v = 0; v < n_nodes; ++v) {
+      ulong uv = u * n_nodes + v;
+      for (ulong k = 0; k < kernel_size; ++k) {
+        ulong vk = v * kernel_size + k;
+        double t = t0 + k * dt + .5 * dt;  // kernel_discretization[k] + .5 * dt;
```
Comment (PhilipDeegan) [inline lib/cpp-test/hawkes/inference/hawkes_em_gtest.cpp:169]: are these helpful in anyway?

Comment (claudio-ICL) [reply lib/cpp-test/hawkes/inference/hawkes_em_gtest.cpp:169]: I think so. The evaluation of the Hawkes kernel is very consequential for the EM algorithm. Before these tests were in place, I could not really be sure about the correctness of the evaluation.

Comment (claudio-ICL) [reply lib/cpp-test/hawkes/inference/hawkes_em_gtest.cpp:169]: ah, sure, that was meant to link the manual increment to the point in the array. I can take that out.

Comment (claudio-ICL) [reply lib/cpp-test/hawkes/inference/hawkes_em_gtest.cpp]: Done here: e43af7e02cd75d01a935728a4fd81342b8f07d77

**Pull request merged**